proposed/local-sdk-global-json.md

```diff
@@ -0,0 +1,112 @@
+- [scripts](https://github.com/dotnet/runtime/blob/main/dotnet.cmd) that wrap `dotnet` to find the  _correct_ `dotnet` to use during build.
+
+These scripts are not one offs, they are increasingly common items in repos in `github.com/dotnet` to attempt to fix the disconnect. Even so many of these solutions are incomplete because they themselves only consider local deployment. They dont't fully support the full set of ways the SDK can be deployed.
+
+This problem also manifests in how customers naturally want to use our development tools like Visual Studio or VS Code. It's felt sharply on the .NET team, or any external customer who wants to contribute to .NET, due to how arcade infrastructure uses xcopy deployment into `.dotnet`. External teams like Unity also feel this pain in their development:
+
+- This [issue](https://github.com/dotnet/sdk/issues/8254) from 2017 attempting to solve this problem. It gets several hits a year from customers who are similarly struggling with our toolings inability to handle local deployment.
+- This [internal discussion](https://teams.microsoft.com/l/message/19:ed7a508bf00c4b088a7760359f0d0308@thread.skype/1698341652961?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=4ba7372f-2799-4677-89f0-7a1aaea3706c&parentMessageId=1698341652961&teamName=.NET%20Developer%20Experience&channelName=InfraSwat&createdTime=1698341652961) from a C# team member. They wanted to use VS as the product is shipped to customers and got blocked when we shipped an SDK that didn't have a corresponding MSI and hence VS couldn't load Roslyn anymore.
+- [VS code](https://github.com/dotnet/vscode-csharp/issues/6471) having to adjust to consider local directories for SDK because our resolver can't find them.
+
+## Detailed Design
+The global.json file will support two new properties under the `sdk` object:
+
+- `"additionalPaths"`: this is a list of paths that the host resolver should consider when looking for compatible SDKs. Relative paths will be interpreted relative to the global.json file.
+- `"additionalPathsOnly"`: when `true` the resolver will _only_ consider paths in `additionalPaths`. It will not consider any machine wide locations (unless they are specified in `additionalPaths`). The default for this property is `false`.
+
+The `additionalPaths` works similar to how multi-level lookup works. It adds additional locations that the host resolver should consider when trying to resolve a compatible .NET SDK. For example:
+
+```json
+{
+    "sdk": {
+        "additionalPaths": [ ".dotnet" ],
+    }
+}
+```
+
+In this configuration the host resolver would find a compatible SDK, if it exists in `.dotnet` or a machine wide location. The host resolver will consider the `additionalPaths` in the order they are defined and will stop at the first match. If none of the locations have a matching SDK then
+the host resolver will fall back to the existing SDK resolution strategy: `%PATH%` followed by machine wide installations.
```

Comment:
> existing SDK resolution strategy: `%PATH%` followed by machine wide installations.

As of .NET 7, the only place the host's resolution logic will look for the SDK is relative to the running `dotnet` executable. That is, it won't look in the machine-wide registered path per Windows registry or `/etc/dotnet/install_location` or the default machine-wide path (`%ProgramFiles%\dotnet`, `/usr/share/dotnet`, `/usr/local/share/dotnet`).

The host also does not use `%PATH%` (and I'd rather it stay that way). Since `dotnet` may be from `%PATH%`, the host may appear to be affected, but it is just based on the running `dotnet`.

Reply:
I think I may be missing the distinction between the host and the resolver here. The key is for the core resolution to respect the new design such that products like VS and VS Code are effectively ignorant of this change. They just ask the resolver for the SDK as they've always done and the fact that we have a new strategy for finding it is an implementation detail. 

@rainersigwald, @baronfel as they pushed me in the current direction.